src/components/application_manager/include/application_manager/application_data_impl.h

```diff
@@ -330,7 +330,7 @@ class DynamicApplicationDataImpl : public virtual Application {
   CommandsMap commands_;
   mutable std::shared_ptr<sync_primitives::RecursiveLock> commands_lock_ptr_;
   SubMenuMap sub_menu_;
-  mutable std::shared_ptr<sync_primitives::Lock> sub_menu_lock_ptr_;
+  mutable std::shared_ptr<sync_primitives::RecursiveLock> sub_menu_lock_ptr_;
```

Comment:
It looks to me that this change is not necessary since the data accessor is only used at the root level.

Reply:
I believe it is necessary, i got a deadlock without that change on this [line](https://github.com/smartdevicelink/sdl_core/pull/3447/files/1e59f31353831c4316ed3a70c161ac5611e47a12..6923723764dfc3514e1305a73972c0121e6bef97#diff-5ab8eae8ebc07c54021736d8a81015a8R118)

The first lock taken is [here](https://github.com/smartdevicelink/sdl_core/pull/3447/files/1e59f31353831c4316ed3a70c161ac5611e47a12..6923723764dfc3514e1305a73972c0121e6bef97#diff-5ab8eae8ebc07c54021736d8a81015a8R214).

Note that the commands accessor has a recursive lock for the same type of usage [here](https://github.com/smartdevicelink/sdl_core/pull/3447/files/1e59f31353831c4316ed3a70c161ac5611e47a12..6923723764dfc3514e1305a73972c0121e6bef97#diff-5ab8eae8ebc07c54021736d8a81015a8R175)

Reply:
Ah, I see I didn't notice that that the lock was taken in the `app->RemoveSubMenu` call